File: dataask-frontend/lib/api/workspaces.ts
/**
 * Workspace API client
 */

import { apiClient } from './client'

export interface Workspace {
  id: string
  name: string
  description?: string
  owner_id: string
  created_at: string
  updated_at: string
}

export interface CreateWorkspaceData {
  name: string
  description?: string
}

export interface UpdateWorkspaceData {
  name?: string
  description?: string
}

export const workspacesApi = {
  /**
   * Get all workspaces for the current user
   */
  async list(): Promise<Workspace[]> {
    const response = await apiClient.get<Workspace[]>('/workspaces')
    return response.data
  },

  /**
   * Get a specific workspace by ID
   */
  async get(id: string): Promise<Workspace> {
    const response = await apiClient.get<Workspace>(`/workspaces/${id}`)
    return response.data
  },

  /**
   * Create a new workspace
   */
  async create(data: CreateWorkspaceData): Promise<Workspace> {
    const response = await apiClient.post<Workspace>('/workspaces', data)
    return response.data
  },

  /**
   * Update a workspace
   */
  async update(id: string, data: UpdateWorkspaceData): Promise<Workspace> {
    const response = await apiClient.patch<Workspace>(`/workspaces/${id}`, data)
    return response.data
  },

  /**
   * Delete a workspace
   */
  async delete(id: string): Promise<void> {
    await apiClient.delete(`/workspaces/${id}`)
  },
}

File: dataask-backend/app/api/routers/workspaces.py
"""Workspace API endpoints."""
import uuid

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.models.workspace import Workspace, WorkspaceMember
from app.schemas.workspace import (
    WorkspaceCreate,
    WorkspaceResponse,
    WorkspaceUpdate,
    WorkspaceMemberCreate,
    WorkspaceMemberResponse,
)

router = APIRouter()


# Temporary: Hard-coded user ID for demo (will be replaced with auth)
DEMO_USER_ID = uuid.uuid4()


@router.get("/workspaces", response_model=list[WorkspaceResponse])
async def list_workspaces(db: AsyncSession = Depends(get_db)) -> list[Workspace]:
    """List all workspaces for the current user."""
    result = await db.execute(
        select(Workspace).where(Workspace.owner_id == DEMO_USER_ID)
    )
    workspaces = result.scalars().all()
    return list(workspaces)


@router.post("/workspaces", response_model=WorkspaceResponse, status_code=status.HTTP_201_CREATED)
async def create_workspace(
    workspace: WorkspaceCreate,
    db: AsyncSession = Depends(get_db),
) -> Workspace:
    """Create a new workspace."""
    db_workspace = Workspace(
        name=workspace.name,
        description=workspace.description,
        settings=workspace.settings,
        owner_id=DEMO_USER_ID,
    )
    db.add(db_workspace)
    await db.commit()
    await db.refresh(db_workspace)
    return db_workspace


@router.get("/workspaces/{workspace_id}", response_model=WorkspaceResponse)
async def get_workspace(
    workspace_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
) -> Workspace:
    """Get a workspace by ID."""
    result = await db.execute(
        select(Workspace).where(Workspace.id == workspace_id)
    )
    workspace = result.scalar_one_or_none()
    if not workspace:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Workspace not found",
        )
    return workspace


@router.patch("/workspaces/{workspace_id}", response_model=WorkspaceResponse)
async def update_workspace(
    workspace_id: uuid.UUID,
    workspace_update: WorkspaceUpdate,
    db: AsyncSession = Depends(get_db),
) -> Workspace:
    """Update a workspace."""
    result = await db.execute(
        select(Workspace).where(Workspace.id == workspace_id)
    )
    workspace = result.scalar_one_or_none()
    if not workspace:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Workspace not found",
        )

    # Update fields
    if workspace_update.name is not None:
        workspace.name = workspace_update.name
    if workspace_update.description is not None:
        workspace.description = workspace_update.description
    if workspace_update.settings is not None:
        workspace.settings = workspace_update.settings

    await db.commit()
    await db.refresh(workspace)
    return workspace


@router.delete("/workspaces/{workspace_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_workspace(
    workspace_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
) -> None:
    """Delete a workspace."""
    result = await db.execute(
        select(Workspace).where(Workspace.id == workspace_id)
    )
    workspace = result.scalar_one_or_none()
    if not workspace:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Workspace not found",
        )

    await db.delete(workspace)
    await db.commit()
